llvm/lib/Passes/StandardInstrumentations.cpp

```diff
@@ -830,6 +831,182 @@ void PrintIRInstrumentation::registerCallbacks(
   }
 }
 
+void DumpIRInstrumentation::registerCallbacks(
+    PassInstrumentationCallbacks &PIC) {
+
+  if (!(shouldDumpBeforeSomePass() || shouldDumpAfterSomePass()))
+    return;
+
+  this->PIC = &PIC;
+
+  PIC.registerBeforeNonSkippedPassCallback(
+      [this](StringRef P, Any IR) { this->pushPass(P, IR); });
+
+  if (shouldDumpBeforeSomePass())
+    PIC.registerBeforeNonSkippedPassCallback(
+        [this](StringRef P, Any IR) { this->dumpBeforePass(P, IR); });
+
+  if (shouldDumpAfterSomePass()) {
+    PIC.registerAfterPassCallback(
+        [this](StringRef P, Any IR, const PreservedAnalyses &) {
+          this->dumpAfterPass(P, IR);
+        });
+  }
+
+  // It is important the the "popPass" callback fires after the dumpAfterPass
+  // callback
+  PIC.registerAfterPassCallback(
+      [this](StringRef P, Any IR, const PreservedAnalyses &) {
+        this->popPass(P);
+      });
+}
+
+void DumpIRInstrumentation::dumpBeforePass(StringRef PassID, Any IR) {
+  if (isIgnored(PassID))
```

Comment:
I think this function doesn't do any effect, maybe just leave it empty?

Reply:
Which function do you mean?

Reply:
> Which function do you mean?

`DumpIRInstrumentation::dumpBeforePass`. Since both `isIgnored` and `shouldDumpBeforePass` do not have any side effect, I believe `DumpIRInstrumentation::dumpBeforePass` basically does nothing.

Reply:
What about these lines:

```
SmallString<16> OutputPath =
      fetchCurrentInstrumentationDumpFile("-before.ll");
  std::error_code EC;
  llvm::raw_fd_ostream OS(OutputPath, EC, llvm::sys::fs::CD_CreateAlways);
  if (EC)
    report_fatal_error(Twine("Failed to open ") + OutputPath +
                       " to support dump-before: " + EC.message());

  OS << "*** IR Dump Before " << PassID << " on " << getIRName(IR) << " ***\n";
  unwrapAndPrint(OS, IR);
```

This method is responsible for printing the IR to a file before a pass runs.

Reply:
somehow I thought the brace on line 870 marks the end of this function, my bad. It's all good now